[Frontend] Disaggregate prefill decode with zmq

panf2333 opened the pull request:
Added vLLM Connect to initiate a proxy service and connect to the VLLM Server via ZMQ, improved the performance of prefill-decode disaggregation by 5-10% (TTFT), and 5- 15% (ITL) on average. 

This key change of this PR includes replacing HTTP with ZMQ for communication between the proxy and the VLLM server, and using socket pools to maintain persistent ZMQ connections, which reduces reconnection overhead.

We have attached the benchmark result and the detailed configuration to reproduce the result.

## Benchmark

Without CUDA_LAUNCH_BLOCKING=1. The TTFT performance is improved by an average of 5-10%, The average performance improvement of ITL is 5-15%

### Parameters
- GPU device: 2 * H100 80G
- Model: meta-llama/Meta-Llama-3.1-8B-Instruct
- Parameters: gpu-memory-utilization 0.6 + kv_buffer_size 5e9
- dataset input 1024 output 6
- QPS: 12, 24, 48, 96
- Total Request: 96

### Evaluation Steps
1. Start Disagg HTTP proxy and 2 VLLM Server Instances(1 prefill and 1 decode)
2. Run the script to test QPS in [12, 24, 48, 96] each qps repeat 3 times and then obtain the average metric
3. Start Disagg ZMQ proxy and 2 VLLM Server Instance and repeat the previous process.
    
![image](https://github.com/user-attachments/assets/08c73d09-82ba-4273-a969-64df7a841e7e)


When I set CUDA_LAUNCH_BLOCKING=1 the results were similar to before,  improved the performance of prefill-decode disaggregation by 30-50% (TTFT), and 3X- 15X (ITL) on average.

### Parameters
- GPU device: 2 * H100 80G
- Model: meta-llama/Meta-Llama-3.1-8B-Instruct
- Parameters: gpu-memory-utilization 0.6 + kv_buffer_size 5e9
- dataset input 1024 output 6
- CUDA_LAUNCH_BLOCKING=1
- QPS: 12, 24, 48, 96
- Total Request: 96

### Evaluation Steps
1. Start Disagg HTTP proxy and 2 VLLM Server Instances(1 prefill and 1 decode)
2. Run the script to test QPS in [12, 24, 48, 96] each qps repeat 3 times and then obtain the average metric
3. Start Disagg ZMQ proxy and 2 VLLM Server Instance and repeat the previous process.
    
![image](https://github.com/user-attachments/assets/9c2e8c4f-1687-4511-b3e6-263ac91da725)




## Design of ZMQ-based Client-Server Communication
### High-level Overview
![image](https://github.com/user-attachments/assets/9d3dc998-1e90-41c5-acce-89581471980d)

### Design of ZMQ-based Communication
![image](https://github.com/user-attachments/assets/8a2e4f6d-900f-4805-9303-ac7dad1f76a9)

panf2333 committed:
add test py

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
add vllm connect cmd

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
add identity url headers

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
add  /v1/completions stream support

Signed-off-by: clark <panf2333@gmail.com>

KuntaiDu commented on vllm/entrypoints/connect.py:
```diff
@@ -0,0 +1,128 @@
+import json
```
The filename `connect` is a big vague --- maybe `disagg_connector` or some other names that are a bit more concrete

panf2333 replied:
Good idea, I will follow it

panf2333 replied:
Done

KuntaiDu commented on vllm/scripts.py:
```diff
@@ -192,6 +196,18 @@ def main():
+        "connect",
+        help="Start the vLLM OpenAI Compatible API server And Connect to other servers disaggreate prefill and decode",
+        usage="vllm connect <model_tag> [options]")
+    connect_parser.add_argument("--prefill-addr",
```
Maybe make this argument `required`.

panf2333 replied:
done

> Maybe make this argument `required`.

KuntaiDu commented on vllm/entrypoints/connect.py:
```diff
@@ -0,0 +1,128 @@
+        prefill_request['max_tokens'] = 1
+        route = "/v1/completions"
+        # finish prefill
+        async for x in execute_task_async(route, header, prefill_request, app.state.sockets_prefill):
```
A potential optimization (you don't need to implement it in this PR): return the first token generated by the prefill instance in this `async for`, instead of reposting the request to decode instance and waiting for the first token from there.

KuntaiDu commented on vllm/scripts.py:
```diff
@@ -192,6 +196,18 @@ def main():
+    connect_parser.add_argument("--prefill-addr",
+                              type=str,
+                              help="The prefill address IP:PORT")
+    connect_parser.add_argument("--decode-addr",
```
DITTO

panf2333 replied:
done

> DITTO

panf2333 commented on vllm/entrypoints/launcher.py:
```diff
@@ -55,8 +59,39 @@ async def dummy_shutdown() -> None:
+    clients.bind(url_client)
+    logger.info(f"ZMQ Server ROUTER started at {url_client}")
+    # Socket to talk to workers
+    workers = context.socket(zmq.DEALER)
```
https://zguide.zeromq.org/docs/chapter3/#The-DEALER-to-DEALER-Combination

We need to proactively send messages to workers in this scenario.

1. ROUTER is not suitable for initiating messages because it doesn't know the identities of other receivers until it receives the first message. Only then can it establish routes for interaction.

2. REQ requires acknowledging each message before sending the next one, which doesn't meet our requirements. 

3. **DEALER allows us to actively send messages and supports asynchronous multi-send and multi-receive, making it the more suitable pattern. It's important to note that we need to maintain the DEALER's ID.**

robertgshaw2-neuralmagic commented on benchmarks/disagg_benchmarks/zmq/test_connect_server1.py:
```diff
@@ -0,0 +1,71 @@
+    print("Worker DEALER started at", url_worker)
+
+    tasks = [asyncio.create_task(worker_routine(url_worker, context, i)) for i in range(5)]
+    proxy_task =  asyncio.to_thread(zmq.proxy, clients, workers)
```
zmq sockets are not threadsafe. This cannot run in a background thread it must be in an asyncio task.

robertgshaw2-neuralmagic replied:
This means you cannot use the built in proxy since it does not use async sockets. In prior versions of VLLM, you will have to write your own proxy

panf2333 replied:
The test scripts test_connect_server1.py and test_connect_server2.py were used to simulate model responses. I've since removed them.

robertgshaw2-neuralmagic commented on vllm/entrypoints/connect.py:
```diff
@@ -0,0 +1,128 @@
+    yield
+    ## close zmq context
+    logger.info("term zmqctx")
+    await app.state.zmqctx.term()
```
use `destroy(linger=0)`

panf2333 replied:
Great point! To ensure immediate termination and avoid potential blocking, I'll switch to using destroy(linger=0) instead of term(). I also replace it in the vllm/entrypoints/launcher.py

https://pyzmq.readthedocs.io/en/latest/api/zmq.html#context
After interrupting all blocking calls, term shall block until the following conditions are satisfied:
1. All sockets open within context have been closed.
2. For each socket within context, all messages sent on the socket have either been physically transferred to a network peer, or the socket’s linger period set with the zmq.LINGER socket option has expired.

robertgshaw2-neuralmagic commented on vllm/entrypoints/launcher.py:
```diff
@@ -55,8 +59,39 @@ async def dummy_shutdown() -> None:
+    logger.info(f"ZMQ Worker DEALER started at {url_worker}")
+
+    tasks = [asyncio.create_task(worker_routine(url_worker, app, context, i)) for i in range(5)]
+    proxy_task =  asyncio.to_thread(zmq.proxy, clients, workers)
```
zmq sockets are not threadsafe. You cannot run this in a background thread.

panf2333 replied:
I appreciate you pointing out potential thread safety issues with zmq sockets. You are completely correct; By default, they are not thread safe. I will prioritize finding a more thread safe alternative in the future to ensure robust operation in multi-threaded environments.

As zmq.proxy() is a synchronous function, executing it directly within the main thread can potentially block the server.

Currently, these two sockets are used exclusively within this thread. While I believe there are no immediate thread safety concerns, it's prudent to consider future scalability and maintainability.  Can we address potential thread-safety issues in a subsequent PR?


https://zguide.zeromq.org/docs/chapter2/#ZeroMQ-s-Built-In-Proxy-Function
It’s exactly like starting the main loop of rrbroker.
![image](https://github.com/user-attachments/assets/c9725134-488a-482b-bcac-48d1255a720d)

https://github.com/booksbyus/zguide/blob/master/examples/Python/rrbroker.py

panf2333 replied:
@robertgshaw2-neuralmagic Hi, Robert I have resolved this issue by [threadproxy](https://pyzmq.readthedocs.io/en/latest/api/zmq.devices.html#threadproxy). It will create sockets in the proxy, which no one can access. So there won't be any thread safety issues

robertgshaw2-neuralmagic commented:
@panf2333 - I am responsible for the multiprocessing + asyncio + zmq architecture of VLLM, so I am going to review this in detail. I am having some trouble following the design here. Can you make a simple diagram that charts out what these objects are to ease in review?

Thanks!

panf2333 committed:
1. connect_parser set --prefill-addr and --decode-addr  are required
2.To more accurately reflect its purpose, we will rename connect.py to disagg_connector.py.

Signed-off-by: clark <panf2333@gmail.com>

panf2333 commented:
> @panf2333 - Thanks for the PR! Disaggregated serving is a hugely important initiative for VLLM in 2025
> 
> I am responsible for the multiprocessing + asyncio + zmq architecture of VLLM, so I am going to review this in detail. I am having some trouble following the design here. Can you make a simple diagram that charts out what these objects are to ease in review?
> 
> Thanks!

It's my pleasure. I'll put together a diagram and send it over shortly.

panf2333 commented:
> @panf2333 - Thanks for the PR! Disaggregated serving is a hugely important initiative for VLLM in 2025
> 
> I am responsible for the multiprocessing + asyncio + zmq architecture of VLLM, so I am going to review this in detail. I am having some trouble following the design here. Can you make a simple diagram that charts out what these objects are to ease in review?
> 
> Thanks!

@robertgshaw2-neuralmagic These are simple diagram , hoping to help you better understand this PR. I also updated the description of PR.

### The relationship with client connector and vllm server
![image](https://github.com/user-attachments/assets/9d3dc998-1e90-41c5-acce-89581471980d)

### The zmq detail between connector and vllm server
![image](https://github.com/user-attachments/assets/8a2e4f6d-900f-4805-9303-ac7dad1f76a9)

panf2333 committed:
update disagg_connect test_request.py

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
Replace zmq.asyncio.Context().term() with zmq.asyncio.Context().destroy(linger=0) for immediate termination

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
1. fix mypy issue

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
Run yapf and ruff

Signed-off-by: clark <panf2333@gmail.com>

russellb reviewed (changes requested):
I don't understand the whole design yet, but I have one early comment: is all zmq communication local? If so, can you please use `ipc://` sockets instead of `tcp://`? That will avoid some security concerns.

panf2333 commented:
> I don't understand the whole design yet, but I have one early comment: is all zmq communication local? If so, can you please use `ipc://` sockets instead of `tcp://`? That will avoid some security concerns.
@russellb I completely agree that security is a paramount concern.

Given the Disaggregated serving feature's potential to dispatch requests to other nodes, it's crucial to establish a secure communication channel between the connector proxy, prefill node, and decode node.

In order to connect the connector proxy, pre filled nodes, and decoding nodes, we should use 'tcp://'

in vllm/entrypoints/disagg_connector.py
`async def run_disagg_connector(args, **uvicorn_kwargs) -> None:`

in vllm/entrypoints/launcher.py
```
async def serve_zmq(arg, zmq_server_port: int, app: FastAPI) -> None:
    """Server routine"""
    logger.info("zmq Server start arg: %s, zmq_server_port: %d", arg,
                zmq_server_port)
    url_worker = "inproc://workers"
    url_client = f"tcp://0.0.0.0:{zmq_server_port}"
```

In the server side we will use "inproc://workers" to deal the message.

russellb commented:
This is big and complex enough that I would find it easier to discuss this at a design doc level. Do you have a design doc from planning this implementation?

I'm not really comfortable with adding any additional multi-node zmq usage without additional non-trivial effort to secure these communications.

noooop commented:
I think a service discovery is needed first. Avoid hardcoding all ports

panf2333 commented:
> This is big and complex enough that I would find it easier to discuss this at a design doc level. Do you have a design doc from planning this implementation?
> 
> I'm not really comfortable with adding any additional multi-node zmq usage without additional non-trivial effort to secure these communications.

@russellb I appreciate you raising this concern. 
I will integrate the pyzme.auth module to enhance security  with follow-up pr. I will change to `ipc://` this time.
> https://pyzmq.readthedocs.io/en/latest/api/zmq.auth.html
> base on ZAP authentication and CURVE authentication 
The document are here. Recommended Lark Doc.


lark doc: https://qus2es1bg99i.larksuite.com/wiki/Pbi1wFUTaiBZneksfytuQxrSsTe?from=from_copylink

google doc: https://docs.google.com/document/d/1ZwFij2OEx_K1xBx2EBx5FKfXQ9EJEGU6shYh-9MJdPs/edit?usp=sharing

panf2333 committed:
1. replace tpc:// with ipc:// \n 2. fix json response

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
create proxy sockets in the proxy function for thread safety

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
fix ThreadProxy

Signed-off-by: clark <panf2333@gmail.com>

panf2333 commented:
> This is big and complex enough that I would find it easier to discuss this at a design doc level. Do you have a design doc from planning this implementation?
> 
> I'm not really comfortable with adding any additional multi-node zmq usage without additional non-trivial effort to secure these communications.

@russellb Hi Russell, for now, I've used 'ipc://' to address immediate security concerns. However, I'll be addressing network security comprehensively in a future PR. I plan to leverage[ pyzmq.auth ](https://pyzmq.readthedocs.io/en/latest/api/zmq.auth.html)to implement robust authentication and authorization mechanisms.

panf2333 committed:
run format.sh

Signed-off-by: clark <panf2333@gmail.com>

russellb commented:
> > This is big and complex enough that I would find it easier to discuss this at a design doc level. Do you have a design doc from planning this implementation?
> > I'm not really comfortable with adding any additional multi-node zmq usage without additional non-trivial effort to secure these communications.
> 
> @russellb Hi Russell, for now, I've used 'ipc://' to address immediate security concerns. However, I'll be addressing network security comprehensively in a future PR. I plan to leverage[ pyzmq.auth ](https://pyzmq.readthedocs.io/en/latest/api/zmq.auth.html)to implement robust authentication and authorization mechanisms.

I don't think that's sufficient. We also need a viable option for encryption, ideally with TLS.

panf2333 commented:
> > > This is big and complex enough that I would find it easier to discuss this at a design doc level. Do you have a design doc from planning this implementation?
> > > I'm not really comfortable with adding any additional multi-node zmq usage without additional non-trivial effort to secure these communications.
> > 
> > 
> > @russellb Hi Russell, for now, I've used 'ipc://' to address immediate security concerns. However, I'll be addressing network security comprehensively in a future PR. I plan to leverage[ pyzmq.auth ](https://pyzmq.readthedocs.io/en/latest/api/zmq.auth.html)to implement robust authentication and authorization mechanisms.
> 
> I don't think that's sufficient. We also need a viable option for encryption, ideally with TLS.

@russellb I believe the disaggregation feature might benefit from optional TLS encryption. While encryption enhances security, it may introduce a slight performance overhead.  Do you mean we can provide a configuration option to enable TLS encryption? This will allow users to choose the security level they need. I think users prefer to deploy clusters within secure environments such as intranets, so they want to improve performance as much as possible.

I will conduct in-depth research on auth and encryption before deciding on the selection. Before that zmq was only allowed to run locally. How about this?
﻿
﻿
﻿

russellb commented:
That's fine. I'm completely OK with using it local-only.

panf2333 committed:
add benchmark shell

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
run format

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
remove invalid zmq benchmark code

Signed-off-by: clark <panf2333@gmail.com>

KuntaiDu reviewed (changes requested):
Great work! Can you change the disaggregated prefill example file under the `examples` folder? Let's provide some handle for newcomers to run disaggregated prefill example without figuring out how to correctly set all the CLI args.

panf2333 committed:
change disagg_prefill example to use zmq

Signed-off-by: clark <panf2333@gmail.com>

panf2333 commented:
> Great work! Can you change the disaggregated prefill example file under the `examples` folder? Let's provide some handle for newcomers to run disaggregated prefill example without figuring out how to correctly set all the CLI args.

done

KuntaiDu reviewed (dismissed):
LGTM!

KuntaiDu commented:
@robertgshaw2-redhat would be great if you can take a look, if it also looks good to you I'll enable automerge.

robertgshaw2-redhat commented on vllm/entrypoints/launcher.py:
```diff
@@ -58,6 +62,37 @@ async def dummy_shutdown() -> None:
+        ]
+        logger.info("zmq tasks: %s", tasks)
+        # thread safety proxy create socket in the background:
+        # https://pyzmq.readthedocs.io/en/latest/api/zmq.devices.html#proxy-devices
```
- How are you dealing with HWM? (https://zeromq.org/get-started/)
- `DEALER` and `ROUTER` sockets will drop messages if they hit HWM

panf2333 replied:
fixed by disable
```
# unlimited HWM
hwm_limit = 0
thread_proxy.bind_in(clients_addr)
thread_proxy.setsockopt_in(zmq.SNDHWM, hwm_limit)
thread_proxy.setsockopt_in(zmq.RCVHWM, hwm_limit)
thread_proxy.bind_out(workers_addr)
thread_proxy.setsockopt_out(zmq.SNDHWM, hwm_limit)
thread_proxy.setsockopt_out(zmq.RCVHWM, hwm_limit)
```

robertgshaw2-redhat commented on vllm/entrypoints/launcher.py:
```diff
@@ -58,6 +62,37 @@ async def dummy_shutdown() -> None:
+    context = zmq.asyncio.Context.instance()
+    try:
+        tasks = [
+            asyncio.create_task(worker_routine(workers_addr, app, context, i))
```
- Why do we need to have 100 workers?
- Since each worker is a coroutine anyways, couldn't we just have one worker that dispatches work?

robertgshaw2-redhat commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -799,6 +799,12 @@ def signal_handler(*_) -> None:
         model_config = await engine_client.get_model_config()
         await init_app_state(engine_client, model_config, app.state, args)
 
+        zmq_server_port = args.zmq_server_port
```
Am I understanding this right that this launches both?
- the zmq server
- the http openai server

I dont understand why this should happen. Shouldn't it be one of the other?

robertgshaw2-redhat replied:
I think a better entrypoint is not to put this under `vllm serve` but rather to have another entrypoint called `vllm xxx`

panf2333 replied:
Yes, we will now start two servers. In my opinion, the zmq server is based on the vllm server. We can reuse the HTTP server functionality. We only replace the means of transportation

I will discuss this with @KuntaiDu

robertgshaw2-redhat replied:
But here we launch both the `zmq` server and the `http` server. Also this is confusing for maintainers of the API server.

Please setup the zmq server as a separate entrypoint. It can basically reuse everything in this function.

robertgshaw2-redhat commented on vllm/entrypoints/disagg_connector.py:
```diff
@@ -0,0 +1,234 @@
+
+    uvloop.run(run_disagg_connector(args))
+
+    # uvicorn.run(app, host="0.0.0.0", port=fastapi_port)
```
?

panf2333 replied:
@robertgshaw2-redhat Do you mean to delete the comment? I have added the main function here for easy debugging
I deleted the comment line.

robertgshaw2-redhat replied:
yes

robertgshaw2-redhat commented on vllm/entrypoints/disagg_connector.py:
```diff
@@ -0,0 +1,234 @@
+        original_request_data = await request.json()
+        logger.info("Received request: %s header: %s", original_request_data,
+                    header)
+        prefill_request = original_request_data.copy()
```
I think we should avoid copying the whole object and instead just temporarily modify `max_tokens`

panf2333 replied:
Fixed

robertgshaw2-redhat commented on vllm/entrypoints/disagg_connector.py:
```diff
@@ -0,0 +1,234 @@
+    # create socket pool with prefill and decode
+    logger.info("start create_socket_pool")
+    app.state.zmqctx = zmq.asyncio.Context()
+    app.state.sockets_prefill = await create_socket_pool(
```
What is the benefit of having a pool of sockets here rather than just having one?

robertgshaw2-redhat replied:
Oh I see, its so that each request "owns" its own socket during the lifecycle of the request.

Doesn't this imply that we can only have `num_socket` active requests at a time?

robertgshaw2-redhat replied:
A better design for this might look like the following:
- have an `output_handler_loop` coroutine that runs in the background, pulling a single socket workers
- the `output_handler_loop` inspects the outputs from the socket and pushes the outputs into a per request asyncio queue for processing by `execute_task_async`

This is basically the design we have for `AsyncLLM` in V1 of VLLM which allows us to have a single socket pair despite having N coroutines running at once.

- https://github.com/vllm-project/vllm/pull/9826

robertgshaw2-redhat replied:
Yeah. I really do not think we should have a design of N sockets for N maximum connections. E.g. this would prevent any batch workloads from running in VLLM.

Since we only have two processes, I think that we should just have a (PUSH/PULL) socket pair and all communications should proceed this way. The design of `AsyncLLM` is setup to deal with the fact that there are N coroutines using the same socket for communication, which is exactly the situation we have here.

I came to the design of `AsyncLLM` after many iterations over 6+ mo of using zmq inside VLLM. This is also similar to how Sglang works internally. It would take a lot of convincing for me to accept this PR with a pattern that uses a separate socket per request. For instance, you are probably going to go over the `ulimit` on ubuntu 😨 

I haven't reviewed the worker yet in detail, but looking at it briefly, I suspect we will want to rework it to fit into this pattern as well.

panf2333 replied:
@robertgshaw2-redhat  this is[ my design](https://qus2es1bg99i.larksuite.com/wiki/NN9DwrzYxiDC41kGVyYuYx96sPe?from=from_copylink), please review it. Because we want to disconnect from multiple nodes later. So DEALER and ROUTER maybe more reasonable.
![image](https://github.com/user-attachments/assets/d8d02053-0c4f-40be-aa09-4016c748f30d)

panf2333 replied:
@robertgshaw2-redhat Hi, Could you please take some time to review my design? Any feedback would be greatly appreciated.

robertgshaw2-redhat commented on vllm/entrypoints/disagg_connector.py:
```diff
@@ -0,0 +1,234 @@
+             headersJson.encode(),
+             requestBody.encode()]),
+                               timeout=time_out)
+        logger.info("Sent end")
```
We cannot have logs like this in production code.

These should be `logger.debug` at most (and probably removed).

robertgshaw2-redhat commented on vllm/entrypoints/disagg_connector.py:
```diff
@@ -0,0 +1,234 @@
+from vllm.logger import init_logger
+from vllm.utils import FlexibleArgumentParser
+
+# default prefill and decode addr
```
Most of these should be arguments to file. We should not have hardcoded ports, especially since vllm serve allows the zmq ports to be specified

panf2333 replied:
These addresses will be passed through the parameters in vllm connect, and they are mandatory, so these few are useless. I will delete them.

robertgshaw2-redhat commented on vllm/entrypoints/disagg_connector.py:
```diff
@@ -0,0 +1,234 @@
+# default prefill and decode addr
+time_out = 180
+fastapi_port = 8000
+prefill_addr = "ipc://localhost:7010"
```
When we specific `ipc` like this, does it go through the local filesystem or over http?
- cc @russellb

russellb replied:
My guess is that this was previously a `tcp://` socket, but was changed at my request to not use TCP for this.

By only changing the prefix to `ipc://`, my guess is that this is creating a UNIX socket with a filename of `localhost:7010` in the current working directory.

panf2333 replied:
Yes, this will create a socket file
![image](https://github.com/user-attachments/assets/f71d982f-6f18-4cfc-af18-af7a8d8013fb)

robertgshaw2-redhat replied:
Makes sense. Probably better to have a different name since its confusing to have a url like object as a file.

panf2333 committed:
change log level and fix some comments

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
remove default socket address value

Signed-off-by: clark <panf2333@gmail.com>

robertgshaw2-redhat commented:
Also QQ - are you planning to maintain this moving forward?

panf2333 commented:
> Also QQ - are you planning to maintain this moving forward?

Of course, Russell has mentioned the security issues of using TCP, so IPC is currently being used. In the future, it is necessary to implement prefill and decode between multiple nodes in a secure environment

panf2333 committed:
fix SIM105

Signed-off-by: clark <panf2333@gmail.com>

KuntaiDu commented:
BTW @panf2333 would be great if you can merge latest vLLM's code and redo the benchmark (IIRC there is a critical bug fix for disagg prefill).

robertgshaw2-redhat commented on vllm/entrypoints/disagg_connector.py:
```diff
@@ -0,0 +1,228 @@
+        await sockets.put(sock)
+
+
+async def generate_stream_response(fisrt_reply: str,
```
typo

robertgshaw2-redhat commented on vllm/entrypoints/disagg_connector.py:
```diff
@@ -0,0 +1,228 @@
+            prefill_response = await prefill(route, header, request_data)
+            if isinstance(prefill_response, JSONResponse):
+                return prefill_response
+            logger.info("finish prefill start decode")
```
debug logs

robertgshaw2-redhat reviewed (changes requested):
Hey guys, nice progress so far. 

My experience with ZMQ inside vLLM over the past year tells me that we that we need to reduce the number of sockets that are being created and used. ZMQ is a great framework, but IPC will be a frequent cause of bugs, especially at high scale - you dont want to spend time debugging why ZMQ is going mute or dropping messages :).

### `serve_zmq` / `worker_routine`

Currently `serve_zmq`:
- creates 100 `worker_routine` coroutines
- runs a thread proxy to route requests from `ipc://127.0.0.1:{zmq_server_port}` to each worker via ipc.

Having a fixed number of worker co-routines is an anti-pattern for asyncio and puts a hard cap on the number of active requests that can be processed concurrently. Additionally, using ZMQ for IPC between two coroutines adds an avenue for bugs and adds overhead without much benefit.

I think we can just remove the `THREAD_PROXY`  and instead do the following:

```python
async def serve_zmq(...):
    context = zmq.asyncio.Context()
    socket = zmq.asyncio.socket(zmq.DEALER)
    running_requests: Set[asyncio.Task] = set()

    while True:
        # get new request from the client
        identity, req = await socket.recv_multipart()
           
         # launch request handler coroutine
         task = asyncio.create_task(worker_routine(identity, req, socket))
         running_requests.add(task)
         task.add_done_callback(running_requests.discard)

async def worker_routine(identity, req, socket):
    # process request - this is pseudocode
    response = process_request(req)
      
    # send back via the DEALER
    await socket.send_multipart(identity, response)
```

### `disagg_connector`

Currently `disagg_connector`,
- creates 100 decode sockets in a queue
- creates 100 prefill sockets in a queue
- when a request arrives, it pops one of the sockets from prefill, sends a prefill req and waits for response. It then pops one of the sockets from decode, sends a decode req and waits for response

Weaknesses of this design:
- we are limited to a fixed number of active requests in the system, which is not ideal
- we have 100s of sockets in the system, which is bug prone

Instead, we should adopt a pattern that we have used across vLLM which is 1 socket or 1 socket pair per communication path. I understand that we may have N prefill worker nodes and M decode worker nodes, but we should have 1 socket for each of these connections. This will keep the system simple and ensure we can scale the number of coroutines using said socket arbitraryily
- We can always provide more resources to the zmq socket by setting `io_threads > 1`

The following design enables this:
- create a request_queues objects (used for communication b/t coroutines)
- create prefill_handler and decode_handler background coroutines, which pull from the sockets and push into the request_queues by inspecting the request_id
- /chat/completion creates a request_queue for the specific request, passes it to prefill and decode which wait on this queue rather than waiting on the socket itself. the prefill_handler and decode_handler a responsible for pulling from the socket and putting into this queue

```python

# communication between output handlers and execute_task_async
request_queues: Dict[str, asyncio.Queue]
ctx = zmq.asyncio.Context()

prefill_socket = ctx.Socket(path_to_prefill)
decode_socket = ctx.Socket(path_to_decode)

# you can have N of these if there are N prefill workers
async def prefill_handler(request):
     while True:
          response = await prefill_socket.recv_multipart()
          request_queues[response.req_id].put_nowait(response)

# you can have M of these if there are M decode workers
async def decode_handler(request):
     while True:
          response = await decode_socket.recv_multipart()
          request_queues[response.req_id].put_nowait(response)

async def execute_task_async(socket, request, queue):
      # send to the socket
      await socket.send_multipart(request)
      
      # handlers will put the response from the socket into this queue
      return await queue.get()

async def prefill(request, queue):
     response = await execute_task_async(prefill_socket, request, queue)
     ...

async def decode(request, queue):
     response = await execute_task_async(decode_socket, request, queue)
     ...

@app.post('/v1/completions')
async def chat_completions(request: Request):
     # create new queue for communication between output handlers and execute_task_async
     request_id = uuid()
     queue = asyncio.Queue()
     request_queues[request_id] = queue

     ...
     prefill_response = await prefill(request, queue)
     ...
     decode_response = await decode(request, queue)

async def run_disagg_connector()
    # launch the prefill_handler and decode_handler coroutines here
    ...
```

This is the design we use for `AsyncLLM` and it works very well

panf2333 commented:
> BTW @panf2333 would be great if you can merge latest vLLM's code and redo the benchmark (IIRC there is a critical bug fix for disagg prefill).

Okay, after refactoring the code, I will perform benchmark testing again.

panf2333 commented:
> ```python
> prefill
> ```

panf2333 committed:
refactor disagg

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
refactor zmq msg to object

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
fix mypy

Signed-off-by: clark <panf2333@gmail.com>

panf2333 committed:
add unlimited HWM

Signed-off-by: clark <panf2333@gmail.com>

panf2333 commented:
When I set CUDA_LAUNCH_BLOCKING=1 the results were similar to before,  improved the performance of prefill-decode disaggregation by 30-50% (TTFT), and 3X- 15X (ITL) on average.

Parameters

    GPU device: 2 * H100 80G
    Model: meta-llama/Meta-Llama-3.1-8B-Instruct
    Parameters: gpu-memory-utilization 0.6 + kv_buffer_size 5e9
    dataset input 1024 output 6
    CUDA_LAUNCH_BLOCKING=1
    QPS: 12, 24, 48, 96
    Total Request: 96

Evaluation Steps

    1. Start Disagg HTTP proxy and 2 VLLM Server Instances(1 prefill and 1 decode)
    2. Run the script to test QPS in [12, 24, 48, 96] each qps repeat 3 times and then obtain the average metric
    3. Start Disagg ZMQ proxy and 2 VLLM Server Instance and repeat the previous process.
    
    
![image](https://github.com/user-attachments/assets/9c2e8c4f-1687-4511-b3e6-263ac91da725)


And When I don't set CUDA_LAUNCH_BLOCKING=1. The performance improvement is not that significant.TTFT performance is improved by an average of 5-10%, The average performance improvement of ITL is 5-15%


Parameters

    GPU device: 2 * H100 80G
    Model: meta-llama/Meta-Llama-3.1-8B-Instruct
    Parameters: gpu-memory-utilization 0.6 + kv_buffer_size 5e9
    dataset input 1024 output 6
    QPS: 12, 24, 48, 96
    Total Request: 96

Evaluation Steps

    1. Start Disagg HTTP proxy and 2 VLLM Server Instances(1 prefill and 1 decode)
    2. Run the script to test QPS in [12, 24, 48, 96] each qps repeat 3 times and then obtain the average metric
    3. Start Disagg ZMQ proxy and 2 VLLM Server Instance and repeat the previous process.

![image](https://github.com/user-attachments/assets/08c73d09-82ba-4273-a969-64df7a841e7e)

panf2333 committed:
fix rebase

Signed-off-by: clark <panf2333@gmail.com>

Commit:
updated

Signed-off-by: Robert Shaw <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: Robert Shaw <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: Robert Shaw <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: Robert Shaw <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: Robert Shaw <rshaw@neuralmagic.com>

Commit:
Stash

Signed-off-by: Robert Shaw <rshaw@neuralmagic.com>

Commit:
Merge branch 'main' into rob-fixes

Commit:
updated

Signed-off-by: Robert Shaw <rshaw@neuralmagic.com>

Commit:
working?

Signed-off-by: Robert Shaw <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: Robert Shaw <rshaw@neuralmagic.com>

Commit:
cleanup

Signed-off-by: Robert Shaw <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: Robert Shaw <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: Robert Shaw <rshaw@neuralmagic.com>

Commit:
fix pre-commit

Signed-off-by: Robert Shaw <rshaw@neuralmagic.com>

robertgshaw2-redhat committed:
pre-commit

Signed-off-by: rshaw@neuralmagic.com <robertgshaw2@gmail.com>

robertgshaw2-redhat committed:
updated

Signed-off-by: rshaw@neuralmagic.com <robertgshaw2@gmail.com>

robertgshaw2-redhat committed:
added files

Signed-off-by: rshaw@neuralmagic.com <robertgshaw2@gmail.com>

robertgshaw2-redhat committed:
updated

Signed-off-by: rshaw@neuralmagic.com <robertgshaw2@gmail.com>

robertgshaw2-redhat committed:
updated

Signed-off-by: rshaw@neuralmagic.com <robertgshaw2@gmail.com>

robertgshaw2-redhat committed:
added __init__.py

Signed-off-by: rshaw@neuralmagic.com <robertgshaw2@gmail.com>

robertgshaw2-redhat committed:
added __init__.py

Signed-off-by: rshaw@neuralmagic.com <robertgshaw2@gmail.com>

robertgshaw2-redhat committed:
updated

Signed-off-by: rshaw@neuralmagic.com <robertgshaw2@gmail.com>

robertgshaw2-redhat committed:
updated

Signed-off-by: rshaw@neuralmagic.com <robertgshaw2@gmail.com>

robertgshaw2-redhat committed:
updated

Signed-off-by: rshaw@neuralmagic.com <robertgshaw2@gmail.com>

robertgshaw2-redhat committed:
updated

Signed-off-by: rshaw@neuralmagic.com <robertgshaw2@gmail.com>

panf2333 committed:
Merge pull request #2 from robertgshaw2-redhat/rob-pd-controller

Rob pd controller
